### DISCUSSION TITLE
Terminal -> package

### DISCUSSION THREAD

**Discussion by Moosems** [tomlin7/biscuit#85] [General]
Could you make a package similar to cupcake for the terminal that uses a TTK theme (you can choose the styling options ;) )? Also, I haven't been near my computer to test Biscuit but can you change the directory in the terminal?

**Comment by tomlin7** [comment]
Yess that's a really good idea ❤ But I'm worried that the current terminal isn't that optimized and is a really poor solution. When I profiled Biscuit, the most critical area was the terminal manager xd. Two threads are spawned per each terminal instance to read from stdout and stderr, so it's really a heavy component. The current terminal manager looks like this.

![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/03a9fccd-045d-4c1e-b868-6531709ebc45)

- can spawn multiple instances
- can be switched between them with the tabs and close them individually

I will look forward into optimizing the terminal solution and then if its a fairly good one, publish it as a standalone embeddable widget!

**Reply by Moosems** [reply]
The terminal looks just like the one in VS Code :D!

**Reply by Moosems** [reply]
Can you change the directory?

**Reply by tomlin7** [reply]
Yess! that is possible with `cd` command
![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/50e1dde0-101c-482e-860a-b8a9f5feb1d0)

**Reply by tomlin7** [reply]
Also you just reminded me that there should be a way to change directory from the interface! 
and a major bug too: the terminal should be opened at the opened folder's directory, currently its opened at the app directory xd

**Reply by Moosems** [reply]
Would you mind if I borrowed some of the logic for another project called TkTerm?

**Reply by tomlin7** [reply]
No problem, you can surely use the code! I'd love to receive some suggestions if you were able to optimize it further too ❤

**Reply by Moosems** [reply]
If you start running an instance of Biscuit from the Biscuit terminal and then close the original Biscuit, does the second one also close or does the Thread persist?

**Reply by Moosems** [reply]
If you purposefully put a correct and incorrect command in the same line, does it successfully run the good commands and give errors on the broken one?

Example:
`ls; cd nonexistent_folder; pwd; cd ..; pwd`

**Reply by Moosems** [reply]
Could you compare the output of that with a normal terminal?

**Reply by tomlin7** [reply]
setting the thread to be a daemon will do that for you, when the main thread terminates, these will too then!

**Reply by tomlin7** [reply]
see here!
https://github.com/billyeatcookies/Biscuit/blob/main/biscuit/core/components/views/panel/terminal/terminal.py#L65-L66

**Reply by tomlin7** [reply]
the Terminal widget takes a mandatory `shell` argument, also it inherits from views.panelview, but you can simply replace that to `tk.Frame`

**Reply by tomlin7** [reply]
@Moosems apparently that wasn't true, the terminal do support that. windows cmd doesn't support running multiple commands in that format!

![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/3b998c58-10af-4976-ae59-1e25c96c3bb2)

**Reply by Moosems** [reply]
Maybe you should change the Mac commands to the respective Windows ones ;).

**Reply by tomlin7** [reply]
@Moosems it supports running multiple commands yes!
here's a comparison
![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/c382c6d3-2ee0-4f23-96d6-02af56ead4a8)
![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/195e30ba-d0ed-467a-9340-fd9f038adca4)

**Reply by Moosems** [reply]
Can you run a valid command after an invalid one?

**Reply by tomlin7** [reply]
@Moosems yes you can, the terminal doesnt crash on command failing

**Reply by Moosems** [reply]
Beautiful :).

**Comment by Moosems** [comment]
Does the terminal support visual apps like nano and vim?

**Reply by tomlin7** [reply]
@Moosems I have looked into this issue before, it's not supported, same goes for python repl, or any application that enters a loop of its own, and i have not been able to fix it

![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/87aa5f75-5b18-42ef-941f-789e519d8a4e)

**Reply by Moosems** [reply]
If I run a python app that calls for input, will it work?

**Reply by tomlin7** [reply]
@Moosems yes that works fine
```py
a = input("enter name: ")
print(a)
```
![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/6edb600c-0260-44d4-ae85-3c0719c26111)

**Reply by Moosems** [reply]
I didn't see it print David :/.

**Reply by Moosems** [reply]
https://github.com/marc2332/tauri-terminal

**Reply by tomlin7** [reply]
@Moosems sorry my bad! xd
but have a look at this
```py
a = input("enter name: ")
print(a)
print(a)
```
![image](https://github.com/billyeatcookies/Biscuit/assets/70792552/28960ea3-ac4d-4878-9a8a-038da2e7092f)

**Reply by Moosems** [reply]
Interesting. Something to look into for sure.

**Reply by Moosems** [reply]
https://github.com/spyder-ide/spyder-terminal

**Reply by Moosems** [reply]
https://github.com/xtermjs/xterm.js

**Reply by tomlin7** [reply]
@Moosems my friend once worked on a terminal emulator, this was the only issue seemed troublesome and later lead to she leaving the project 😅

but yes, these terminals are pretty complicated, and i will look into them, will try researching more

**Comment by cid0rz** [comment]
Terminal buffering is complicated

**Reply by tomlin7** [reply]
yep, and hope we will figure it out one day xd

**Comment by Moosems** [comment]
@billyeatcookies take a look at this: https://wiki.tcl-lang.org/page/terminal+emulator

The Expect extension looks pretty promising. What if you made a wrapper around it?

**Reply by tomlin7** [reply]
@Moosems couldn't really dig up most of the information, also i'm really bad at tcl, and i haven't researched on the wrapper topic any much!

**Reply by Moosems** [reply]
Take some time eventually, I think you'll be happily surprised.

**Comment by Moosems** [comment]
Anything on the separate package idea?

**Reply by tomlin7** [reply]
i rewrote the terminal recently. Now it can run programs that enter a loop too. But as of now, I'm not parsing and using the ANSI codes generated properly yet. I'm just printing plain text after stripping off all the ANSI escapes. 

The separate package idea might help with development here, so yes I'm thinking about it, I'll update here soon.

**Reply by Moosems** [reply]
Nice!

**Comment by tomlin7** [comment]
@Moosems lets continue with this topic at https://github.com/billyeatcookies/mono
